browser/resources/settings/default_brave_shields_page/components/brave_adblock_editor.js

```diff
@@ -0,0 +1,119 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import {I18nMixin} from 'chrome://resources/js/i18n_mixin.js';
+import {PolymerElement, html} from 'chrome://resources/polymer/v3_0/polymer/polymer_bundled.min.js';
+import {BaseMixin} from '../../base_mixin.js';
+import {PrefsMixin} from '../../prefs/prefs_mixin.js';
+
+import 'chrome://resources/cr_elements/cr_button/cr_button.m.js';
+
+const AdBlockEditorBase = PrefsMixin(I18nMixin(BaseMixin(PolymerElement)))
+
+class AdBlockEditor extends AdBlockEditorBase {
+  static get is() {
+    return 'adblock-editor'
+  }
+
+  static get template() {
+    return html`{__html_template__}`
+  }
+
+  static get properties() {
+    return {
+      value: {
+        type: Text,
+        observer: 'lineNumberObserver_'
+      },
+      activeLineNumber_: {
+        type: Number,
+        value: null
+      },
+      lines_: Array,
```

Comment:
Perhaps this would be simplified a little if `lines_` could be a computed property and an object which stored character start index for each line so you didn't need to recompute. It could also store wrap indexes too. At that point the selection logic could be a little simpler